// wikijs/dist/wiki.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _Promise = require('babel-runtime/core-js/promise')['default'];

var _Object$keys = require('babel-runtime/core-js/object/keys')['default'];

var _interopRequireWildcard = require('babel-runtime/helpers/interop-require-wildcard')['default'];

Object.defineProperty(exports, '__esModule', {
	value: true
});

var _request = require('request-promise');

var _request2 = _interopRequireWildcard(_request);

var _import = require('underscore');

var _import2 = _interopRequireWildcard(_import);

var _markupParser = require('./wiki-markup-parser');

var _markupParser2 = _interopRequireWildcard(_markupParser);

'use strict';

/**
 * @namespace
 * @property {string} apiUrl - URL of Wikipedia API
 */
var defaultOptions = {
	apiUrl: 'https://en.wikipedia.org/w/api.php'
};

/**
* Wiki
* @class
* @param  {Object} [options] - override options for API (only apiUrl for now)
* @return {Wiki}
*/

var Wiki = (function () {
	function Wiki(options) {
		_classCallCheck(this, Wiki);

		this.options = _import2['default'].extend(options || {}, defaultOptions);
	}

	_createClass(Wiki, [{
		key: 'api',
		value: function api(params) {
			var _this = this;

			return new _Promise(function (resolve, reject) {
				_request2['default'].get({
					uri: _this.options.apiUrl,
					qs: _import2['default'].extend(params, {
						format: 'json',
						action: 'query'
					}),
					headers: {
						'User-Agent': 'WikiJs/0.1 (https://github.com/dijs/wiki; richard.vanderdys@gmail.com)'
					}
				}).then(function (res) {
					return resolve(JSON.parse(res));
				})['catch'](reject);
			});
		}
	}, {
		key: 'pagination',
		value: function pagination(params, parseResults) {
			var _this2 = this;

			return new _Promise(function (resolve, reject) {
				_this2.api(params).then(function (res) {
					var resolution = {};
					resolution.results = parseResults(res);
					if (res['query-continue']) {
						var type = _Object$keys(res['query-continue'])[0];
						var continueKey = _Object$keys(res['query-continue'][type])[0];
						params[continueKey] = res['query-continue'][type][continueKey];
						resolution.next = function () {
							return _this2.pagination(params, parseResults);
						};
					}
					resolve(resolution);
				})['catch'](reject);
			});
		}
	}, {
		key: 'aggregatePagination',
		value: function aggregatePagination(pagination) {
			var _this3 = this;

			var allResults = arguments[1] === undefined ? [] : arguments[1];

			return new _Promise(function (resolve, reject) {
				pagination.then(function (res) {
					res.results.forEach(function (result) {
						return allResults.push(result);
					});
					if (res.next) {
						resolve(_this3.aggregatePagination(res.next(), allResults));
					} else {
						resolve(allResults);
					}
				})['catch'](reject);
			});
		}
	}, {
		key: 'search',

		/**
   * Search articles
   * @example
   * new Wiki().search('star wars').then((data) => console.log(data.results.length));
   * @example
   * new Wiki().search('star wars').then((data) => {
   * 	data.next().then(...);
   * });
   * @method Wiki#search
   * @param  {string} query - keyword query
   * @param  {Number} [limit] - limits the number of results
   * @return {Promise} - pagination promise with results and next page function
   */
		value: function search(query) {
			var limit = arguments[1] === undefined ? 50 : arguments[1];

			return this.pagination({
				list: 'search',
				srsearch: query,
				srlimit: limit
			}, function (res) {
				return _import2['default'].pluck(res.query.search, 'title');
			});
		}
	}, {
		key: 'random',

		/**
   * Random articles
   * @method Wiki#random
   * @param  {Number} [limit] - limits the number of random articles
   * @return {Promise}
   */
		value: function random() {
			var _this4 = this;

			var limit = arguments[0] === undefined ? 1 : arguments[0];

			return new _Promise(function (resolve, reject) {
				_this4.api({
					list: 'random',
					rnnamespace: 0,
					rnlimit: limit
				}).then(function (res) {
					return resolve(_import2['default'].pluck(res.query.random, 'title'));
				})['catch'](reject);
			});
		}
	}, {
		key: 'page',

		/**
   * Create Page Interface
   * @example
   * new Wiki().page('Batman').then((page) => console.log(page.pageid));
   * @method Wiki#page
   * @param  {string} title - title of article
   * @return {Promise}
   */
		value: function page(title) {
			var _this5 = this;

			return new _Promise(function (resolve, reject) {
				_this5.api({
					prop: 'info|pageprops',
					inprop: 'url',
					ppprop: 'disambiguation',
					titles: title
				}).then(function (res) {
					var id = _import2['default'].findKey(res.query.pages, function (page) {
						return page.title === title;
					});
					if (!id) {
						reject(new Error('No article found'));
					} else {
						resolve(new WikiPage(res.query.pages[id], _this5));
					}
				})['catch'](reject);
			});
		}
	}, {
		key: 'geoSearch',

		/**
   * Geographical Search
   * @method Wiki#geoSearch
   * @param  {Number} lat - latitude
   * @param  {Number} lon - longitude
   * @param  {Number} [radius] - search radius in kilometers
   * @return {Promise}
   */
		value: function geoSearch(lat, lon) {
			var _this6 = this;

			var radius = arguments[2] === undefined ? 1000 : arguments[2];
			// 1km
			return new _Promise(function (resolve, reject) {
				_this6.api({
					list: 'geosearch',
					gsradius: radius,
					gscoord: lat + '|' + lon
				}).then(function (res) {
					return resolve(_import2['default'].pluck(res.query.geosearch, 'title'));
				})['catch'](reject);
			});
		}
	}]);

	return Wiki;
})();

/**
* Page Interface
* @class WikiPage
* @param  {Object} [props] - page properties from API page query
* @return {WikiPage}
*/

var WikiPage = (function () {
	function WikiPage(props, wiki) {
		_classCallCheck(this, WikiPage);

		this.wiki = wiki;
		_import2['default'].extend(this, props);
	}

	_createClass(WikiPage, [{
		key: 'html',

		/**
   * HTML from page
   * @method WikiPage#html
   * @return {Promise}
   */
		value: function html() {
			var _this7 = this;

			return new _Promise(function (resolve, reject) {
				_this7.wiki.api({
					prop: 'revisions',
					rvprop: 'content',
					rvlimit: 1,
					rvparse: '',
					titles: _this7.title
				}).then(function (res) {
					return resolve(res.query.pages[_this7.pageid].revisions[0]['*']);
				})['catch'](reject);
			});
		}
	}, {
		key: 'content',

		/**
   * Text content from page
   * @method WikiPage#content
   * @return {Promise}
   */
		value: function content() {
			var _this8 = this;

			return new _Promise(function (resolve, reject) {
				_this8.wiki.api({
					prop: 'extracts',
					explaintext: '',
					titles: _this8.title
				}).then(function (res) {
					return resolve(res.query.pages[_this8.pageid].extract);
				})['catch'](reject);
			});
		}
	}, {
		key: 'summary',

		/**
   * Text summary from page
   * @method WikiPage#summary
   * @return {Promise}
   */
		value: function summary() {
			var _this9 = this;

			return new _Promise(function (resolve, reject) {
				_this9.wiki.api({
					prop: 'extracts',
					explaintext: '',
					exintro: '',
					titles: _this9.title
				}).then(function (res) {
					return resolve(res.query.pages[_this9.pageid].extract);
				})['catch'](reject);
			});
		}
	}, {
		key: 'images',

		/**
   * Image URL's from page
   * @method WikiPage#images
   * @return {Promise}
   */
		value: function images() {
			var _this10 = this;

			return new _Promise(function (resolve, reject) {
				_this10.wiki.api({
					generator: 'images',
					gimlimit: 'max',
					prop: 'imageinfo',
					iiprop: 'url',
					titles: _this10.title
				}).then(function (res) {
					return resolve(_import2['default'].chain(res.query.pages).pluck('imageinfo').flatten().pluck('url').value());
				})['catch'](reject);
			});
		}
	}, {
		key: 'references',

		/**
   * References from page
   * @method WikiPage#references
   * @return {Promise}
   */
		value: function references() {
			var _this11 = this;

			return new _Promise(function (resolve, reject) {
				_this11.wiki.api({
					prop: 'extlinks',
					ellimit: 'max',
					titles: _this11.title
				}).then(function (res) {
					return resolve(_import2['default'].pluck(res.query.pages[_this11.pageid].extlinks, '*'));
				})['catch'](reject);
			});
		}
	}, {
		key: 'links',

		/**
   * Paginated links from page
   * @method WikiPage#links
   * @param  {Boolean} [aggregated] - return all links (default is true)
   * @param  {Number} [limit] - number of links per page
   * @return {Promise} - includes results [and next function for more results if not aggregated]
   */
		value: function links() {
			var _this12 = this;

			var aggregated = arguments[0] === undefined ? true : arguments[0];
			var limit = arguments[1] === undefined ? 100 : arguments[1];

			var pagination = this.wiki.pagination({
				prop: 'links',
				plnamespace: 0,
				pllimit: limit,
				titles: this.title
			}, function (res) {
				return _import2['default'].pluck(res.query.pages[_this12.pageid].links, 'title');
			});
			if (aggregated) {
				return this.wiki.aggregatePagination(pagination);
			} else {
				return pagination;
			}
		}
	}, {
		key: 'categories',

		/**
   * Paginated categories from page
   * @method WikiPage#categories
   * @param  {Boolean} [aggregated] - return all categories (default is true)
   * @param  {Number} [limit] - number of categories per page
   * @return {Promise} - includes results [and next function for more results if not aggregated]
   */
		value: function categories() {
			var _this13 = this;

			var aggregated = arguments[0] === undefined ? true : arguments[0];
			var limit = arguments[1] === undefined ? 100 : arguments[1];

			var pagination = this.wiki.pagination({
				prop: 'categories',
				pllimit: limit,
				titles: this.title
			}, function (res) {
				return _import2['default'].pluck(res.query.pages[_this13.pageid].categories, 'title');
			});
			if (aggregated) {
				return this.wiki.aggregatePagination(pagination);
			} else {
				return pagination;
			}
		}
	}, {
		key: 'coordinates',

		/**
   * Geographical coordinates from page
   * @method WikiPage#coordinates
   * @return {Promise}
   */
		value: function coordinates() {
			var _this14 = this;

			return new _Promise(function (resolve, reject) {
				_this14.wiki.api({
					prop: 'coordinates',
					titles: _this14.title
				}).then(function (res) {
					return resolve(res.query.pages[_this14.pageid].coordinates[0]);
				})['catch'](reject);
			});
		}
	}, {
		key: 'info',

		/**
   * Get info from page
   * @method WikiPage#info
   * @return {Promise} - info Object contains key/value pairs of infobox data
   */
		value: function info() {
			var _this15 = this;

			return new _Promise(function (resolve, reject) {
				_this15.wiki.api({
					prop: 'revisions',
					rvprop: 'content',
					rvsection: 0,
					titles: _this15.title
				}).then(function (res) {
					return resolve(_markupParser2['default'](res.query.pages[_this15.pageid].revisions[0]['*']));
				})['catch'](reject);
			});
		}
	}, {
		key: 'backlinks',

		/**
   * Paginated backlinks from page
   * @method WikiPage#backlinks
   * @param  {Boolean} [aggregated] - return all backlinks (default is true)
   * @param  {Number} [limit] - number of backlinks per page
   * @return {Promise} - includes results [and next function for more results if not aggregated]
   */
		value: function backlinks() {
			var aggregated = arguments[0] === undefined ? true : arguments[0];
			var limit = arguments[1] === undefined ? 100 : arguments[1];

			var pagination = this.wiki.pagination({
				list: 'backlinks',
				bllimit: limit,
				bltitle: this.title
			}, function (res) {
				return _import2['default'].pluck(res.query.backlinks, 'title');
			});
			if (aggregated) {
				return this.wiki.aggregatePagination(pagination);
			} else {
				return pagination;
			}
		}
	}]);

	return WikiPage;
})();

/**
 * @module Wiki
 */
exports['default'] = Wiki;
module.exports = exports['default'];
